src/alpakatest/plugin-Test1/alpaka/alpakaAlgo1.cc

```diff
@@ -3,101 +3,311 @@
 namespace {
   constexpr unsigned int NUM_VALUES = 1000;
 
+
   struct vectorAdd {
     template <typename T_Acc, typename T_Data>
     ALPAKA_FN_ACC void operator()(
-        T_Acc const &acc, T_Data const *a, T_Data const *b, T_Data *c, unsigned int numElements) const {
-      //uint32_t const gridDimension(alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0u]);
-      uint32_t const blockDimension(alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
-      uint32_t const gridBlockIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u]);
-      uint32_t const blockThreadIdx(alpaka::idx::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
-      uint32_t const elemDimension(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
-
-      uint32_t i = (blockThreadIdx + gridBlockIdx * blockDimension) * elemDimension;
-      if (i < numElements) {
-        c[i] = a[i] + b[i];
+				  T_Acc const& acc, const T_Data* __restrict__ a, const T_Data* __restrict__ b, T_Data* __restrict__ c, unsigned int numElements) const {
+
+      // Global thread index in grid
+      const uint32_t threadIdxGlobal(alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+      const uint32_t threadDimension(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
+
+      // Global element index (obviously relevant for CPU only, for GPU, i = threadIndexGlobal only)
+      const uint32_t firstElementIdxGlobal = threadIdxGlobal * threadDimension;
+      const uint32_t lastElementIdxGlobal = threadIdxGlobal * threadDimension + threadDimension - 1;
+
+      for (uint32_t i = firstElementIdxGlobal; i <= lastElementIdxGlobal; ++i) {
+	if (i < numElements) {
+	  c[i] = a[i] + b[i];
+	}
       }
```

Comment:
Yes sure, will do.
I just wanted to have a 1:1 implementation with cudatest at first, to be able to compare performances in the CUDA cases and spot issues:
https://github.com/ghugo83/pixeltrack-standalone/blob/master/src/cudatest/plugin-Test1/gpuAlgo1.cu#L12

Reply:
Ok. Functionally the `std::min(numElements, threadIdxGlobal * threadDimension + threadDimension)` does the same as `if (i < numElements)` in `cudatest`.  In principle the `cudatest` versions could be changed to use a loop as well.

Reply:
Yes totally, the min implementation in std must have a < anyway.
But with the use of min, the range of the loop is smaller, which seems to provide a tiny improvement.
Ok will correct in cudatest as well (in separate PR to avoid mixing too much).

Reply:
Ha in cudatest it is not useful, as there is no loop anyway.
Here in alpakatest, it is interesting because it allows to reduce the range of the loop, leading to a tiny improvement.